datafusion/optimizer/src/single_distinct_to_groupby.rs

```diff
@@ -131,162 +127,163 @@ impl OptimizerRule for SingleDistinctToGroupBy {
+                            // --Aggregate: groupBy=[[test.a + Int32(1) AS group_alias_0, test.c AS alias1]], aggr=[[]] [group_alias_0:Int32, alias1:UInt32]\
+                            // ----TableScan: test [a:UInt32, b:UInt32, c:UInt32]
+                            //
+                            // First aggregate(from bottom) refers to `test.a` column.
+                            // Second aggregate refers to the `group_alias_0` column, Which is a valid field in the first aggregate.
+                            // If we were to write plan above as below without alias
+                            //
+                            // Aggregate: groupBy=[[test.a + Int32(1)]], aggr=[[COUNT(alias1)]] [group_alias_0:Int32, COUNT(alias1):Int64;N]\
+                            // --Aggregate: groupBy=[[test.a + Int32(1), test.c AS alias1]], aggr=[[]] [group_alias_0:Int32, alias1:UInt32]\
+                            // ----TableScan: test [a:UInt32, b:UInt32, c:UInt32]
+                            //
+                            // Second aggregate refers to the `test.a + Int32(1)` expression However, its input do not have `test.a` expression in it.
+                            let alias_str = format!("group_alias_{i}");
+                            let (qualifier, field) = schema.qualified_field(i);
+                            (
+                                group_expr.alias(alias_str.clone()),
                                 (
-                                    alias_expr,
-                                    (
-                                        col(alias_str),
-                                        Some(qualified_name(qualifier, field.name())),
-                                    ),
-                                )
-                            }
-                        })
-                        .unzip();
-
-                    // and they can be referenced by the alias in the outer aggr plan
-                    let outer_group_exprs = out_group_expr_with_alias
-                        .iter()
-                        .map(|(out_group_expr, _)| out_group_expr.clone())
-                        .collect::<Vec<_>>();
-
-                    // replace the distinct arg with alias
-                    let mut index = 1;
-                    let mut group_fields_set = HashSet::new();
-                    let mut inner_aggr_exprs = vec![];
-                    let outer_aggr_exprs = aggr_expr
-                        .iter()
-                        .map(|aggr_expr| match aggr_expr {
-                            Expr::AggregateFunction(AggregateFunction {
-                                func_def: AggregateFunctionDefinition::BuiltIn(fun),
-                                args,
-                                distinct,
-                                ..
-                            }) => {
-                                // is_single_distinct_agg ensure args.len=1
-                                if *distinct
```

Comment:
old code had redundant conditionals so I refactored it into a single `if ... else ...`